precompiles/balances-erc20/src/tests.rs

```diff
@@ -0,0 +1,648 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::assert_matches::assert_matches;
+
+use crate::mock::*;
+use crate::*;
+
+use pallet_evm::PrecompileSet;
+use precompile_utils::{error, EvmDataWriter, LogsBuilder};
+use sha3::{Digest, Keccak256};
+
+#[test]
+fn selector_less_than_four_bytes() {
+	ExtBuilder::default().build().execute_with(|| {
+		// This selector is only three bytes long when four are required.
+		let bogus_selector = vec![1u8, 2u8, 3u8];
+
+		assert_eq!(
+			Precompiles::<Runtime>::execute(
+				Account::Precompile.into(),
+				&bogus_selector,
+				None,
+				&evm::Context {
+					address: Account::Precompile.into(),
+					caller: Account::Alice.into(),
+					apparent_value: From::from(0),
+				},
+			),
+			Some(Err(error("tried to parse selector out of bounds")))
+		);
+	});
+}
+
+#[test]
+fn no_selector_exists_but_length_is_right() {
+	ExtBuilder::default().build().execute_with(|| {
+		let bogus_selector = vec![1u8, 2u8, 3u8, 4u8];
+
+		assert_eq!(
+			Precompiles::<Runtime>::execute(
+				Account::Precompile.into(),
+				&bogus_selector,
+				None,
+				&evm::Context {
+					address: Account::Precompile.into(),
+					caller: Account::Alice.into(),
+					apparent_value: From::from(0),
+				},
+			),
+			Some(Err(error("unknown selector")))
+		);
+	});
+}
+
+#[test]
+fn selectors() {
+	assert_eq!(u32::from(Action::BalanceOf), 0x70a08231);
+	assert_eq!(u32::from(Action::TotalSupply), 0x18160ddd);
+	assert_eq!(u32::from(Action::Approve), 0x095ea7b3);
+	assert_eq!(u32::from(Action::Allowance), 0xdd62ed3e);
+	assert_eq!(u32::from(Action::Transfer), 0xa9059cbb);
+	assert_eq!(u32::from(Action::TransferFrom), 0x23b872dd);
+
+	assert_eq!(
+		crate::SELECTOR_LOG_TRANSFER,
+		&Keccak256::digest(b"Transfer(address,address,uint256)")[..]
+	);
+
+	assert_eq!(
+		crate::SELECTOR_LOG_APPROVAL,
+		&Keccak256::digest(b"Approval(address,address,uint256)")[..]
+	);
+}
+
+#[test]
+fn get_total_supply() {
```

Comment:
This is a good typical case, but we could add some edge cases too. For example if the total is zero (maybe not that interesting), or if the total is the top of the numeric range the runtime's balance type supports (which is probably still less that the max u256 that solidity supports).